### ISSUE TITLE
hangs on Waiting for domain to get an IP address only on remote libvirt

### ISSUE THREAD

**Issue by crprince** [vagrant-libvirt/vagrant-libvirt#1252]
### Steps to reproduce
1. configure Vagrantfile to connect to remote libvirt
2. vagrant up
3. Hangs on "Waiting for domain to get an IP Address..."

### Expected behaviour

I am running this on a single host.  If I use connection settings for a remote libvirt (loopback) then the up hang on waiting for the management network ip.  If i remove those settings if up's fine.

### Actual behaviour
Hangs on "Waiting for domain to get an IP Address..."
then .... 
      12: from /root/.vagrant.d/gems/2.6.6/gems/vagrant-libvirt-0.4.1/lib/vagrant-libvirt/action/start_domain.rb:359:in `call'
        11: from /opt/vagrant/embedded/gems/2.2.15/gems/vagrant-2.2.15/lib/vagrant/action/warden.rb:48:in `call'
        10: from /root/.vagrant.d/gems/2.6.6/gems/vagrant-libvirt-0.4.1/lib/vagrant-libvirt/action/wait_till_up.rb:37:in `call'
         9: from /root/.vagrant.d/gems/2.6.6/gems/vagrant-libvirt-0.4.1/lib/vagrant-libvirt/util/timer.rb:9:in `time'
         8: from /root/.vagrant.d/gems/2.6.6/gems/vagrant-libvirt-0.4.1/lib/vagrant-libvirt/action/wait_till_up.rb:38:in `block in call'
         7: from /opt/vagrant/embedded/gems/2.2.15/gems/vagrant-2.2.15/lib/vagrant/util/retryable.rb:17:in `retryable'
         6: from /root/.vagrant.d/gems/2.6.6/gems/vagrant-libvirt-0.4.1/lib/vagrant-libvirt/action/wait_till_up.rb:43:in `block (2 levels) in call'
         5: from /root/.vagrant.d/gems/2.6.6/gems/vagrant-libvirt-0.4.1/lib/vagrant-libvirt/driver.rb:102:in `get_domain_ipaddress'
         4: from /root/.vagrant.d/gems/2.6.6/gems/vagrant-libvirt-0.4.1/lib/vagrant-libvirt/driver.rb:149:in `get_ipaddress_from_domain'
         3: from /root/.vagrant.d/gems/2.6.6/gems/fog-core-2.2.3/lib/fog/core/model.rb:72:in `wait_for'
         2: from /root/.vagrant.d/gems/2.6.6/gems/fog-core-2.2.3/lib/fog/core/wait_for.rb:6:in `wait_for'
         1: from /root/.vagrant.d/gems/2.6.6/gems/fog-core-2.2.3/lib/fog/core/wait_for.rb:6:in `loop'
/root/.vagrant.d/gems/2.6.6/gems/fog-core-2.2.3/lib/fog/core/wait_for.rb:9:in `block in wait_for': The specified wait_for timeout (2 seconds) was exceeded (Fog::Errors::TimeoutError)

### System configuration
**OS/Distro version:**: 
centos/7

**Libvirt version:** 
0.4.1

**Output of `vagrant version; vagrant plugin list`:**
[chris@RB20 vm2]$ sudo vagrant plugin list
vagrant-libvirt (0.4.1, global)
[chris@RB20 vm2]$ vagrant version; vagrant plugin list
Installed Version: 2.2.15
Latest Version: 2.2.15

**Attach Output of `VAGRANT_LOG=debug vagrant ... --provider=libvirt >vagrant.log 2>&1`**
```
 INFO global: Vagrant version: 2.2.15
 INFO global: Ruby version: 2.6.6
 INFO global: RubyGems version: 3.0.3
 INFO global: VAGRANT_LOG="debug"
 INFO global: VAGRANT_INSTALLER_VERSION="2"
 INFO global: VAGRANT_INSTALLER_EMBEDDED_DIR="/opt/vagrant/embedded"
 INFO global: VAGRANT_EXECUTABLE="/opt/vagrant/embedded/gems/2.2.15/gems/vagrant-2.2.15/bin/vagrant"
 INFO global: VAGRANT_INSTALLER_ENV="1"
 WARN global: resolv replacement has not been enabled!
DEBUG global: Loading core plugin: /opt/vagrant/embedded/gems/2.2.15/gems/vagrant-2.2.15/plugins/commands/autocomplete/plugin.rb
 INFO manager: Registered plugin: autocomplete command
DEBUG global: Loading core plugin: /opt/vagrant/embedded/gems/2.2.15/gems/vagrant-2.2.15/plugins/commands/box/plugin.rb
 INFO manager: Registered plugin: box command
DEBUG global: Loading core plugin: /opt/vagrant/embedded/gems/2.2.15/gems/vagrant-2.2.15/plugins/commands/cap/plugin.rb
 INFO manager: Registered plugin: cap command
DEBUG global: Loading core plugin: /opt/vagrant/embedded/gems/2.2.15/gems/vagrant-2.2.15/plugins/commands/cloud/plugin.rb
 INFO manager: Registered plugin: vagrant-cloud
DEBUG global: Loading core plugin: /opt/vagrant/embedded/gems/2.2.15/gems/vagrant-2.2.15/plugins/commands/destroy/plugin.rb
 INFO manager: Registered plugin: destroy command
DEBUG global: Loading core plugin: /opt/vagrant/embedded/gems/2.2.15/gems/vagrant-2.2.15/plugins/commands/global-status/plugin.rb
 INFO manager: Registered plugin: global-status command
DEBUG global: Loading core plugin: /opt/vagrant/embedded/gems/2.2.15/gems/vagrant-2.2.15/plugins/commands/halt/plugin.rb
 INFO manager: Registered plugin: halt command
DEBUG global: Loading core plugin: /opt/vagrant/embedded/gems/2.2.15/gems/vagrant-2.2.15/plugins/commands/help/plugin.rb
 INFO manager: Registered plugin: help command
DEBUG global: Loading core plugin: /opt/vagrant/embedded/gems/2.2.15/gems/vagrant-2.2.15/plugins/commands/init/plugin.rb
 INFO manager: Registered plugin: init command
DEBUG global: Loading core plugin: /opt/vagrant/embedded/gems/2.2.15/gems/vagrant-2.2.15/plugins/commands/list-commands/plugin.rb
 INFO manager: Registered plugin: list-commands command
DEBUG global: Loading core plugin: /opt/vagrant/embedded/gems/2.2.15/gems/vagrant-2.2.15/plugins/commands/login/plugin.rb
 INFO manager: Registered plugin: vagrant-login
DEBUG global: Loading core plugin: /opt/vagrant/embedded/gems/2.2.15/gems/vagrant-2.2.15/plugins/commands/package/plugin.rb
 INFO manager: Registered plugin: package command
DEBUG global: Loading core plugin: /opt/vagrant/embedded/gems/2.2.15/gems/vagrant-2.2.15/plugins/commands/plugin/plugin.rb
 INFO manager: Registered plugin: plugin command
DEBUG global: Loading core plugin: /opt/vagrant/embedded/gems/2.2.15/gems/vagrant-2.2.15/plugins/commands/port/plugin.rb
 INFO manager: Registered plugin: port command
DEBUG global: Loading core plugin: /opt/vagrant/embedded/gems/2.2.15/gems/vagrant-2.2.15/plugins/commands/powershell/plugin.rb
 INFO manager: Registered plugin: powershell command
DEBUG global: Loading core plugin: /opt/vagrant/embedded/gems/2.2.15/gems/vagrant-2.2.15/plugins/commands/provider/plugin.rb
 INFO manager: Registered plugin: provider command
DEBUG global: Loading core plugin: /opt/vagrant/embedded/gems/2.2.15/gems/vagrant-2.2.15/plugins/commands/provision/plugin.rb
 INFO manager: Registered plugin: provision command
DEBUG global: Loading core plugin: /opt/vagrant/embedded/gems/2.2.15/gems/vagrant-2.2.15/plugins/commands/push/plugin.rb
 INFO manager: Registered plugin: push command
DEBUG global: Loading core plugin: /opt/vagrant/embedded/gems/2.2.15/gems/vagrant-2.2.15/plugins/commands/rdp/plugin.rb
 INFO manager: Registered plugin: rdp command
DEBUG global: Loading core plugin: /opt/vagrant/embedded/gems/2.2.15/gems/vagrant-2.2.15/plugins/commands/reload/plugin.rb
 INFO manager: Registered plugin: reload command
DEBUG global: Loading core plugin: /opt/vagrant/embedded/gems/2.2.15/gems/vagrant-2.2.15/plugins/commands/resume/plugin.rb
 INFO manager: Registered plugin: resume command
DEBUG global: Loading core plugin: /opt/vagrant/embedded/gems/2.2.15/gems/vagrant-2.2.15/plugins/commands/snapshot/plugin.rb
 INFO manager: Registered plugin: snapshot command
DEBUG global: Loading core plugin: /opt/vagrant/embedded/gems/2.2.15/gems/vagrant-2.2.15/plugins/commands/ssh/plugin.rb
 INFO manager: Registered plugin: ssh command
DEBUG global: Loading core plugin: /opt/vagrant/embedded/gems/2.2.15/gems/vagrant-2.2.15/plugins/commands/ssh_config/plugin.rb
 INFO manager: Registered plugin: ssh-config command
DEBUG global: Loading core plugin: /opt/vagrant/embedded/gems/2.2.15/gems/vagrant-2.2.15/plugins/commands/status/plugin.rb
 INFO manager: Registered plugin: status command
DEBUG global: Loading core plugin: /opt/vagrant/embedded/gems/2.2.15/gems/vagrant-2.2.15/plugins/commands/suspend/plugin.rb
 INFO manager: Registered plugin: suspend command
DEBUG global: Loading core plugin: /opt/vagrant/embedded/gems/2.2.15/gems/vagrant-2.2.15/plugins/commands/up/plugin.rb
 INFO manager: Registered plugin: up command
DEBUG global: Loading core plugin: /opt/vagrant/embedded/gems/2.2.15/gems/vagrant-2.2.15/plugins/commands/upload/plugin.rb
 INFO manager: Registered plugin: upload command
DEBUG global: Loading core plugin: /opt/vagrant/embedded/gems/2.2.15/gems/vagrant-2.2.15/plugins/commands/validate/plugin.rb
 INFO manager: Registered plugin: validate command
DEBUG global: Loading core plugin: /opt/vagrant/embedded/gems/2.2.15/gems/vagrant-2.2.15/plugins/commands/version/plugin.rb
 INFO manager: Registered plugin: version command
DEBUG global: Loading core plugin: /opt/vagrant/embedded/gems/2.2.15/gems/vagrant-2.2.15/plugins/commands/winrm/plugin.rb

```

**A Vagrantfile to reproduce the issue:**
```
Vagrant.configure("2") do |config|

  config.vm.box = "centos/8"
  config.vm.box_check_update = true
  config.vm.hostname="vm2"

   config.vm.provider :libvirt do |libvirt|
      libvirt.driver="qemu"
      libvirt.host="RB20.local"
      libvirt.connect_via_ssh=true
      libvirt.username="root"
      libvirt.cpus = 2
      libvirt.memory = 2048
   end

end
```

**Are you using upstream vagrant package or your distros package?**
 Distro

**Comment by electrofelix** [comment]
Can you share exactly what the Vagrantfile looks like when you remove the loopback, because it's unusual to try to use libvirt as root, often libvirt and qemu may require additional configuration for remote access as root to behave as expected.

If your Vagrantfile looks like the following and works:
```
Vagrant.configure("2") do |config|

  config.vm.box = "centos/8"
  config.vm.box_check_update = true
  config.vm.hostname="vm2"

   config.vm.provider :libvirt do |libvirt|
      libvirt.driver="qemu"
      libvirt.cpus = 2
      libvirt.memory = 2048
   end
end
```

Then it is likely you are bringing up the VM with a different user when you enable to remote loopback. I'd suggest trying the following for the remote connection (skip the user or explicitly set it to your current user):
```
Vagrant.configure("2") do |config|

  config.vm.box = "centos/8"
  config.vm.box_check_update = true
  config.vm.hostname="vm2"

   config.vm.provider :libvirt do |libvirt|
      libvirt.driver="qemu"
      libvirt.host="RB20.local"
      # uncomment the following to explicitly use your username for the connection
      #libvirt.user = ENV['USER']
      libvirt.connect_via_ssh=true
      libvirt.cpus = 2
      libvirt.memory = 2048
   end
end
```

Secondly, if none of any of the settings `connect_via_ssh`, `host`, `username`, `password`, `id_ssh_key_file`, `qemu_use_session`, or `socket` are set, then the code will check to see if `LIBVIRT_DEFAULT_URI` is defined in the environment, similar to how `virsh` will behave. So check what value that has, as if it points to `qemu:///session` which some distros are doing, it does result in different behaviour as it's trying to use the user session. And to match when you use the remote session you'd need to do the following:
```
Vagrant.configure("2") do |config|

  config.vm.box = "centos/8"
  config.vm.box_check_update = true
  config.vm.hostname="vm2"

   config.vm.provider :libvirt do |libvirt|
      libvirt.driver="qemu"
      libvirt.host="RB20.local"
      # uncomment the following to explicitly use your username for the connection
      #libvirt.user = ENV['USER']
      libvirt.connect_via_ssh=true
      libvirt.qemu_use_session=true
      libvirt.cpus = 2
      libvirt.memory = 2048
   end
end
```

Try and attach the entire vagrant.log as a file attachment, as it'll capture information such as the connection URI used to allow seeing what is happening.

**Comment by crprince** [comment]
Thanks for the reply and suggestions.

There is not much more to the Vagrantfile.  Here is a bit more background.  I'm working on moving a bunch of VMs from a Windows / VirtualBox to Linux / libvirt.   The end goal is to have a Jenkins Build box spin-up remote libvirt VMs.

- This is a small lab, 2 physical servers, both running centos/7 and libvirt at the moment.
- Using the virtualization manager in Gnome, can connect from RB10.local to RB20.local which is headless and create and manipulate VMs.  Seems to work fine so far.
- Not much else is setup yet, so I used root for simplicity
- I can create a VM without the remote connection parameters, and add the remote connection to the Vagrantfile, then halt or destroy the VM for example.  So I believe the connection is working.
- sudo sealert -a /var/log/audit/audit.log shows no alerts.

For my vagrant + libvirt experiment, I started with just the most basic Vagrantfile:

Entire Vagrantfile without the remote connection, running on RB20.local / sudo vagrant up
```
# -*- mode: ruby -*-
# vi: set ft=ruby :

Vagrant.configure("2") do |config|
  config.vm.box = "centos/8"
  config.vm.box_check_update = true
  config.vm.hostname="vm1"

  config.vm.provider :libvirt do |libvirt|
      libvirt.driver="qemu"
      libvirt.cpus = 2
      libvirt.memory = 2048
   end
end
```
A VM is created and starts as expected.   

Entire Vagrantfile with remote connection (to itself)
# -*- mode: ruby -*-
# vi: set ft=ruby :

Vagrant.configure("2") do |config|

  config.vm.box = "centos/8"
  config.vm.box_check_update = true
  config.vm.hostname="vm1"

   config.vm.provider :libvirt do |libvirt|
      libvirt.driver="qemu"
      libvirt.host="RB20.local"
      libvirt.connect_via_ssh=true
      libvirt.username="root"
      libvirt.cpus = 2
      libvirt.memory = 2048
   end
end




It seems odd to me that the connection parameters would cause this.  Here are complete logs for both with and without the remote connection.

- [vagrant_remote.log](https://github.com/vagrant-libvirt/vagrant-libvirt/files/6333285/vagrant_remote.log)
- [vagrant_noremote.log](https://github.com/vagrant-libvirt/vagrant-libvirt/files/6333286/vagrant_noremote.log)

Changing to:  
``
libvirt.user = ENV['USER'] 
``
 Still hangs.

added 

``
   puts "USER=" + ENV['USER'].to_s + "\n"
   puts "LIBVIRT_DEFAULT_URI=" + ENV['LIBVIRT_DEFAULT_URI'].to_s + "\n"
``

[chris@RB20 vm1]$ sudo vagrant up
**USER=root
LIBVIRT_DEFAULT_URI=**
Bringing machine 'default' up with 'libvirt' provider...
==> default: Checking if box 'centos/8' version '2011.0' is up to date...
root@rb20.local's password:
==> default: Creating image (snapshot of base box volume).
==> default: Creating domain with the following settings...
==> default:  -- Name:              vm1_default
==> default:  -- Domain type:       qemu
==> default:  -- Cpus:              2
==> default:  -- Feature:           acpi
==> default:  -- Feature:           apic
==> default:  -- Feature:           pae
==> default:  -- Clock offset:      utc
==> default:  -- Memory:            2048M
==> default:  -- Management MAC:
==> default:  -- Loader:
==> default:  -- Nvram:
==> default:  -- Base box:          centos/8
==> default:  -- Storage pool:      default
==> default:  -- Image:             /var/lib/libvirt/images/vm1_default.img (11G)
==> default:  -- Disk driver opts:  cache='default'
==> default:  -- Kernel:
==> default:  -- Initrd:
==> default:  -- Graphics Type:     vnc
==> default:  -- Graphics Port:     -1
==> default:  -- Graphics IP:       127.0.0.1
==> default:  -- Graphics Password: Not defined
==> default:  -- Video Type:        cirrus
==> default:  -- Video VRAM:        9216
==> default:  -- Sound Type:
==> default:  -- Keymap:            en-us
==> default:  -- TPM Backend:       passthrough
==> default:  -- TPM Path:
==> default:  -- INPUT:             type=mouse, bus=ps2
==> default: Creating shared folders metadata...
==> default: Starting domain.
==> default: Waiting for domain to get an IP address...
...hangs...

Not sure what to try next.

**Comment by electrofelix** [comment]
The connection URI looks to be fine in both cases:
* remote: qemu+ssh://root@RB20.local/system?no_verify=1
* local: qemu:///system

So it's not accidentally using the wrong values from what I can see.

Two possibilities, an issue with the connection password, or an issue with retrieving the address. I see the remote approach asks for your password initially, however there may be more than one connection created and in the case of the other connections the password request may be getting swallowed.

I think the next step is to add a helper to read in the password and pass it to the libvirt provider directly, or set up a password-less ssh key pair.

Possibly we should have some way to indicate during configuration handling when to ask for the password and then use https://github.com/hashicorp/vagrant/blob/main/lib/vagrant/ui.rb#L165-L203

In the short term you can add a helper to the top of your Vagrantfile based on https://stackoverflow.com/a/15685054
```
begin
  require 'io/console'
rescue LoadError
end

def get_password(prompt="Password: ")
  print prompt
  STDIN.noecho(&:gets).chomp
end

remote_pass = get_password()

Vagrant.configure("2") do |config|

  config.vm.box = "centos/8"
  config.vm.box_check_update = true
  config.vm.hostname="vm2"

   config.vm.provider :libvirt do |libvirt|
      libvirt.driver="qemu"
      libvirt.host="RB20.local"
      libvirt.connect_via_ssh=true
      libvirt.qemu_use_session=true
      libvirt.password = remote_pass
      libvirt.cpus = 2
      libvirt.memory = 2048
   end
end
```

If you find that it can't access STDIN there is a way of using `read` instead, though if you are using special characters might need to follow the alternative approaches in the linked stackoverflow post.


If that doesn't help, the next step is around using virsh to connect to the remote machine and see what is getting listed as the, the connection URI looks to be fine in both cases, so there may be some subtle difference around calling the API for listing the dhcp leases.

```
virsh -c qemu+ssh://root@RB20.local/system?no_verify=1 net-dhcp-leases vagrant-libvirt
```

The last option is for the network name.

If you run `VAGRANT_LOG=debug vagrant up` with the remote options enabled, look for the line similar to the following:
```
wait_till_up: Searching for IP for MAC address: 52:54:00:33:b3:f8
```

Basically the code should be looking at the dhcp leases that libvirt is handing out for the mac address, in the case of your log out that was 52:54:00:33:b3:f8, but it'll change on each pristine create.

At the moment I think it's more likely auth based, and potentially hidden by some of the rework done that now skips adding an entry to the uri for an ssh keyfile by default unless it is found to exist. When finalizing the configuration, the code should probably check for either a password or ssh keyfile being provided and if neither is, ask for a password.

**Comment by crprince** [comment]
Something i just realized - when i had the password specified in the Vagrantfile  `libvirt.password=*******` it still prompted me for my password.  I had removed it in anticipation of opening an issue and had reduced the Vagrantfile to barebones.  (In the end, it would have been SSH Keys with a non-root user, etc.  but was taking it step-by-step)   The Virtualization Manager GUI sometimes prompts for a password multiple times if i don't have SSH keys set up.

I will try the above and post results (and move on to ssh keys as well)

**Comment by electrofelix** [comment]
That suggests there is something not quite right around configuring of the password, that sounds like despite it being set it's not getting passed along to the libvirt client correctly in fog-libvirt when the connection is being set up.

**Comment by electrofelix** [comment]
@crprince I've just done another recent test and for me it appears to work as expected using my username and password provided.

Not entirely sure what happened, can you confirm whether it's still broken for you using your password?

If not, I suggest that you modify the local install of your plugin after the lines https://github.com/vagrant-libvirt/vagrant-libvirt/blob/1fe5a80516fb940b4232ce4ab8baaf3b98037dc5/lib/vagrant-libvirt/driver.rb#L30-L35

To add:
```
puts "conn_attr: #{conn_attr}"
```

And just check to see if what is being output matches what you expect to be set.

**Comment by electrofelix** [comment]
@crprince I've recently come across a bit more info as to what was going on here. It turns out that when libvirt has polkit support (which is the default these days), local connections will request interactive authentication with the root password, so it never gets to the point of reading the provided username and password from the Vagrantfile.

https://doc.opensuse.org/documentation/leap/virtualization/html/book-virtualization/cha-libvirt-connect.html#sec-libvirt-connect-auth-libvirt-pk

Essentially it looks like unless libvirt is configured to use SASL for auth or connecting via ssh and the user is in the libvirt/libvirtd group (depending on distro), the username/password combo will be ignored. Sorry for taking so long to finally pin down what was going on.

**Comment by crprince** [comment]
That’s an awesome find; thanks so much for sticking with this.    

 

The focus of my project is changing back to Virtualization and Infrastructure soon; having this explanation will be helpful in selling the approach.

 

Thx

Chris
